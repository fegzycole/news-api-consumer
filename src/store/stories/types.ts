export const SET_STORIES = "SET_STORIES";
export const SET_ERROR = "SET_ERROR";
export const SET_FETCHING_STORIES = "SET_FETCHING_STORIES";

export enum StoryCategories {
  TopHeadlines = "Top Headlines",
  Business = "Business",
  Entertainment = "Entertainment",
  General = "General",
  Health = "Health",
  Science = "Science",
  Sports = "Sports",
  Technology = "Technology",
}

export interface StoryData {
  source: {
    id: null | number;
    name: string;
  };
  author: string;
  title: string;
  description: string;
  url: string;
  urlToImage: string;
  publishedAt: Date | string;
  content: string;
}

export interface StoriesState {
  stories: StoryData[];
  fetchingStories: boolean;
  error?: string;
}

export interface AddStories {
  stories: StoryData[];
  clearStories: boolean;
}

export interface SetStories {
  type: typeof SET_STORIES;
  payload: AddStories;
}

export interface SetError {
  type: typeof SET_ERROR;
  payload: string;
}

export interface SetFetchingStories {
  type: typeof SET_FETCHING_STORIES;
  payload: boolean;
}

export type MoviesActionTypes =
  | SetStories
  | SetError
  | SetFetchingStories;
