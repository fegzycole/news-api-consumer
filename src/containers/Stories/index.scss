@import "../../styles/mixins";

.Stories_Container {
  padding: 1rem;

  @include lgScreen {
    padding: 0px;
    width: 800px;
    margin: 1rem auto 0 auto;
  }
}
