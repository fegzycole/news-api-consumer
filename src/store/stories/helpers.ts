import { StoryCategories } from "./types";

export const apiKey = process.env.REACT_APP_API_KEY;
export const apiUrl = "https://newsapi.org/v2/top-headlines";

export const getRequestPath = (
  category: StoryCategories,
  pageNumber: number
): string => {
  let path = `${apiUrl}?apiKey=${apiKey}&pageSize=10&page=${pageNumber}&language=en&sortBy=publishedAt`;

  if (category !== StoryCategories.TopHeadlines) {
    path = `${path}&category=${category.toLowerCase()}`;
  }

  return path;
};

export const getResponseError = (error: any): string => {
  let serverErrResponse = error?.response?.data?.message;

  if (!serverErrResponse) {
    serverErrResponse = error.message;
  }

  return serverErrResponse;
};
