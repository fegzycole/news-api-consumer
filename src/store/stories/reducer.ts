import {
  MoviesActionTypes,
  StoriesState,
  SET_STORIES,
  SET_ERROR,
  SET_FETCHING_STORIES,
} from "./types";

const initialState: StoriesState = {
  stories: [],
  fetchingStories: false,
  error: undefined,
};

const storiesReducer = (
  state: StoriesState = initialState,
  action: MoviesActionTypes
) => {
  switch (action.type) {
    case SET_STORIES:
      return {
        ...state,
        stories: action.payload.clearStories
          ? action.payload.stories
          : state.stories.concat(action.payload.stories),
        fetchingStories: false,
      };
    case SET_ERROR:
      return {
        ...state,
        error: action.payload,
        fetchingStories: false,
      };
    case SET_FETCHING_STORIES:
      return {
        ...state,
        fetchingStories: action.payload,
      };
    default:
      return state;
  }
};

export default storiesReducer;
