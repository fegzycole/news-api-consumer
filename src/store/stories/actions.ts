import { Dispatch } from "react";
import axios from "axios";

import {
  StoryData,
  SetStories,
  SET_STORIES,
  SetError,
  SET_ERROR,
  SetFetchingStories,
  SET_FETCHING_STORIES,
  StoryCategories,
  AddStories,
} from "./types";
import { getResponseError, getRequestPath } from "./helpers";

export const setStories = (payload: AddStories): SetStories => {
  return {
    type: SET_STORIES,
    payload,
  };
};

export const setError = (payload: string): SetError => {
  return {
    type: SET_ERROR,
    payload,
  };
};

export const setFetchingStories = (payload: boolean): SetFetchingStories => {
  return {
    type: SET_FETCHING_STORIES,
    payload,
  };
};

export const getStories = (
  category: StoryCategories,
  pageNumber: number
): ((
  dispatch: Dispatch<{ type: string; payload?: unknown }>
) => Promise<void>) => {
  return async (dispatch) => {
    dispatch(setError(""));
    dispatch(setFetchingStories(true));

    const requestPath = getRequestPath(category, pageNumber);

    try {
      const request = await axios.get(requestPath);

      const responseData = request.data.articles as StoryData[];

      const stories: StoryData[] = responseData as StoryData[];

      const payload: AddStories = {
        stories,
        clearStories: pageNumber === 1,
      };

      return dispatch(setStories(payload));
    } catch (error: any) {
      const renderedError = getResponseError(error);
      return dispatch(setError(renderedError));
    }
  };
};
