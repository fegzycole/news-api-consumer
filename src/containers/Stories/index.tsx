import { v4 as uuidv4 } from "uuid";
import Story from "../../components/Story";

import { StoryCategories, StoryData } from "../../store/stories/types";
import './index.scss';

export interface StoriesProps {
  topStory: StoryData;
  selectedCategory: StoryCategories;
  stories: StoryData[];
}

const Stories: React.FC<StoriesProps> = ({
  topStory,
  selectedCategory,
  stories,
}) => {
  return (
    <section className="Stories_Container">
      <Story story={topStory} selectedCategory={selectedCategory} topStory />
      {stories.map((story, idx) => (
        <Story
          key={uuidv4()}
          story={story}
          selectedCategory={selectedCategory}
          reverse={idx % 2 === 0}
        />
      ))}
    </section>
  );
};

export default Stories;
