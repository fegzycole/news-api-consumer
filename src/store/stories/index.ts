import * as Types from './types';
import storiesReducer from './reducer';
import * as Actions from './actions';

export { Types, storiesReducer, Actions };
