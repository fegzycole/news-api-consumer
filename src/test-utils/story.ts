import { StoryData } from "../store/stories/types";

const generateStories = () => {
  const mockStories: StoryData[] = [];

  let i = 0;

  while (i < 12) {
    mockStories.push(mockStory);

    i++;
  }

  return mockStories;
};

export const mockStory: StoryData = {
  publishedAt: "Mock Date",
  author: "Mock Author",
  urlToImage: "MockUrlToImage",
  description: "Mock Description",
  source: {
    name: "Mock Source Name",
    id: 1,
  },
  title: "Mock Title",
  url: "mockUrl",
  content: "MockContent",
};

export const mockStories: StoryData[] = generateStories();
