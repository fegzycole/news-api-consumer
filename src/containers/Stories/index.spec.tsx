import Stories, { StoriesProps } from "./index";
import { StoryCategories } from "../../store/stories/types";
import { mockStory } from "../../test-utils/story";
import { render } from "@testing-library/react";

const mockProps: StoriesProps = {
  stories: [mockStory],
  topStory: mockStory,
  selectedCategory: StoryCategories.Sports,
};

jest.mock("../../components/Story", () => ({
  default: () => <div className="Story" />,
  __esModule: true,
  Story: true,
}));

describe("<Stories />", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should match the Stories component snapshot", () => {
    const view = render(<Stories {...mockProps} />);
    expect(view.container).toMatchSnapshot();
  });
});
