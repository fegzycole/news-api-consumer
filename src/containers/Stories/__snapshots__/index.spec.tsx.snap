// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Stories /> should match the Stories component snapshot 1`] = `
<div>
  <section
    class="Stories_Container"
  >
    <div
      class="Story"
    />
    <div
      class="Story"
    />
  </section>
</div>
`;
